Editor keyboard shortcuts

martindzejky opened the pull request:
### What?

Hello guys, so I decided I will try to implement the editor shortcuts. I know that there's a card for that in the [Trello Roadmap board](https://trello.com/c/9eX7KPGy), which kind of makes me think that this is only for the official devs to implement. Is it? If it is, I won't waste time trying, but if it isn't (anyone can implement it), it is currently a lot of fun for me to work on Godot, so I'll do my best.
### How I plan to do it

So I have read the code on how it works in the `Project Settings / Input Map`, how the `InputMap` class works, how `DefaultInput` works, how controls currently receive keyboard events and how this is all set up in the editor. I also now understand that the input map is currently stored in `Global` and saved to engine.cfg. Now as far as I understand it, `EditorSettings` is used to store, well, editor settings.

So my plan is to refactor the existing input map to a separate control, make it work not just with `Global` but also with `EditorSettings` (to store the shortcuts there), then make controls also accept action names for shortcuts (accelerators in code) and respond to them. So this would create a new tab in the editor settings, Shortcuts, which would work the same as Input Map in project settings. Controls would also accept action names (and draw the first assigned key).

This could ease up the creation of new editor shortcuts and potentially let the user set up their own actions and assign them to various buttons and menus in the editor.
### Checklist
- [x] Refactor `ProjectSettings`'s Input Map tab into a separate class
- [x] Make it work with `Global` and `EditorSettings`
- [x] Add the shortcuts tab to editor settings
- [ ] Make controls, menus and buttons accept action name as an accelerator (shortcut)
- [ ] Make controls draw the first assigned key to action
- [ ] Test
- [ ] Profit
- [ ] ???
### The question

Should I work on this? Or let someone else handle this? For now I'll assume that my effort won't be wasted and work on this.

_Note: Don't merge this PR yet... Work in progress..._

martindzejky committed:
Add shortcuts tab to the editor settings dialog

chuckeles commented:
I need help :laughing: I can't get my custom control working so I have to ask. When you add a new class (control), what everything you have to set up for Godot to register it? I want to use signals and I also don't want to show it in the editor.

martindzejky committed:
Add InputEditor

I made this by copying the ProjectSettings and getting rid of the redundant methods and members.
The rest is the same code.

martindzejky committed:
Removed input things from ProjectSettings

martindzejky committed:
Clear action name text after adding

martindzejky committed:
Make InputEditor work with Globals and EditorSettings

martindzejky committed:
Add InputEditor tab to editor settings

martindzejky committed:
Create message popup

Otherwise the editor crashes...

martindzejky committed:
Fix InputEditor not working when not given UndoRedo

martindzejky committed:
Disable adding actions in the editor

martindzejky committed:
Add some scene shortcuts

chuckeles commented:
OK, this is as far as I've got so far, but I'm stuck. I save the shortcuts into the EditorSettings. This works as long as the editor is opened, it even saves properly into the xml config file. But as soon as the editor is started again (and it loads from the file I guess), it overrides my settings with arrays containing only nils! I can't figure this out!

chuckeles commented:
Contents of the config file before the editor is opened:

```xml
...
<array name="input/scene_new" len="1" shared="false">
	<input_event> Device 0 ID 0 Event: Key Unicode:  Scan: 78 Echo: False PressedFalse Mod: C </input_event>
</array>
<array name="input/scene_open" len="1" shared="false">
	<input_event> Device 0 ID 0 Event: Key Unicode:  Scan: 79 Echo: False PressedFalse Mod: C </input_event>
</array>
<array name="input/scene_save" len="1" shared="false">
	<input_event> Device 0 ID 0 Event: Key Unicode:  Scan: 83 Echo: False PressedFalse Mod: C </input_event>
</array>
<array name="input/scene_save_as" len="1" shared="false">
	<input_event> Device 0 ID 0 Event: Key Unicode:  Scan: 83 Echo: False PressedFalse Mod: SC </input_event>
</array>
<array name="input/scene_close" len="1" shared="false">
	<input_event> Device 0 ID 0 Event: Key Unicode:  Scan: 87 Echo: False PressedFalse Mod: C </input_event>
</array>
...
```

Then I start the editor and it becomes:

```xml
...
<array name="input/scene_new" len="1" shared="false">
	<nil>  </nil>
</array>
<array name="input/scene_open" len="1" shared="false">
	<nil>  </nil>
</array>
<array name="input/scene_save" len="1" shared="false">
	<nil>  </nil>
</array>
<array name="input/scene_save_as" len="1" shared="false">
	<nil>  </nil>
</array>
<array name="input/scene_close" len="1" shared="false">
	<nil>  </nil>
</array>
...
```

Anyone from the devs @reduz @akien-mga and others, do you have any idea why this might me occuring? I'm settings this in the `_load_defaults` of the `EditorSettings`:

```cpp
// shortcuts
Array arr;
InputEvent ie;
ie.type = InputEvent::KEY;

ie.key.mod.control = true;
ie.key.scancode = KEY_N;
arr.push_back(ie);
set("input/scene_new", arr);
```
